Mathlib/Tactic/NormNum/BigOperators.lean

```diff
@@ -0,0 +1,384 @@
+/-
+Copyright (c) 2023 Anne Baanen. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Anne Baanen, Floris van Doorn
+
+! This file was ported from Lean 3 source module algebra.big_operators.norm_num
+! leanprover-community/mathlib commit 70fd9563a21e7b963887c9360bd29b2393e6225a
+! Please do not edit these lines, except to modify the commit id
+! if you have ported upstream changes.
+-/
+import Mathlib.Tactic.NormNum.Basic
+import Mathlib.Algebra.BigOperators.Basic
+import Mathlib.Data.List.FinRange
+
+/-!
+# `norm_num` plugin for big operators
+
+This file adds `norm_num` plugins for `Finset.prod` and `Finset.sum`.
+
+The driving part of this plugin is `Mathlib.Meta.NormNum.evalFinsetBigop`.
+We repeatedly use `Finset.proveEmptyOrCons` to try to find a proof that the given set is empty,
+or that it consists of one element inserted into a strict subset, and evaluate the big operator
+on that subset until the set is completely exhausted.
+
+## See also
+
+ * The `fin_cases` tactic has similar scope: splitting out a finite collection into its elements.
+
+## Porting notes
+
+This plugin is noticeably less powerful than the equivalent version in Mathlib 3: the design of
+`norm_num` means plugins have to return numerals, rather than a generic expression.
+In particular, we can't use the plugin on sums containing variables.
+(See also the TODO note "To support variables".)
+
+## TODOs
+
+ * Support intervals: `Finset.Ico`, `Finset.Icc`, ...
+ * To support variables, like in Mathlib 3, turn this into a standalone tactic that unfolds
+   the sum/prod, without computing its numeric value (using the `ring` tactic to do some
+   normalization?)
+-/
+
+namespace Mathlib.Meta
+
+open Lean hiding Rat mkRat
+open Meta
+open Qq
+
+/-- This represents the result of trying to determine whether the given expression `n : Q(ℕ)`
+is either `zero` or `succ`. -/
+inductive Nat.UnifyZeroOrSuccResult (n : Q(ℕ))
+  /-- `n` unifies with `0` -/
+  | zero (pf : $n =Q 0)
+  /-- `n` unifies with `succ n'` for this specific `n'` -/
+  | succ (n' : Q(ℕ)) (pf : $n =Q Nat.succ $n')
+
+/-- Determine whether the expression `n : Q(ℕ)` unifies with `0` or `Nat.succ n'`.
+
+Fails if neither of the options succeed.
+-/
+def Nat.unifyZeroOrSucc (n : Q(ℕ)) : MetaM (Nat.UnifyZeroOrSuccResult n) := do
```

Comment:
The reason I don't use `deriveNat` within the procedure is that we really want definitional equality on the `n` here, since it occurs in types (specifically `Fin n`). I'll try using your described approach for `Finset.range`...